### PR TITLE
`SkinnedMeshRenderer` support immediate initialization of `rootBone` and `bones`

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Feat:

- Remove glTF lazy initialization issue, we can get `rootBone` at anytime
- `SkinnedMeshRenderer` support custom `bones`
- Remove hack logic of Initialize bone data

### CODE REVIEW

File: packages/core/src/mesh/SkinnedMeshRenderer.ts [hunk 1]
```diff
@@ -173,10 +182,10 @@ export class SkinnedMeshRenderer extends MeshRenderer {
     const blendShapeManager = mesh._blendShapeManager;
     blendShapeManager._updateShaderData(shaderData, this);
 
-    const skin = this._skin;
-    if (skin) {
+    const bones = this.bones;
```
Comment (zhuxudong) [inline packages/core/src/mesh/SkinnedMeshRenderer.ts:185]: this._bones

File: packages/core/src/mesh/SkinnedMeshRenderer.ts [hunk 2]
```diff
@@ -391,4 +321,44 @@ export class SkinnedMeshRenderer extends MeshRenderer {
...
+      entity = parent;
+    }
+    return true;
+  }
+
+  /**
+   * @internal
+   */
+  private _getEntityByHierarchyPath(rootEntity: Entity, inversePath: number[]): Entity {
+    let sprite = rootEntity;
```
Comment (zhuxudong) [inline packages/core/src/mesh/SkinnedMeshRenderer.ts:344]: sprite ?

File: packages/core/src/mesh/SkinnedMeshRenderer.ts [hunk 3]
```diff
@@ -391,4 +321,44 @@ export class SkinnedMeshRenderer extends MeshRenderer {
...
+  /**
+   * @deprecated
+   * Skin Object.
+   */
+  get skin(): Skin {
+    return this._skin;
+  }
+
+  set skin(value: Skin) {
+    this._skin = value;
```
Comment (zhuxudong) [inline packages/core/src/mesh/SkinnedMeshRenderer.ts:362]: you should set renderer's `bones` and `rootBond` if you want adapter deprecated method

File: packages/core/src/mesh/SkinnedMeshRenderer.ts [hunk 4]
```diff
@@ -391,4 +321,44 @@ export class SkinnedMeshRenderer extends MeshRenderer {
...
+      const parent = entity.parent;
+      if (parent) {
+        inversePath.push(entity.siblingIndex);
+      } else {
+        return false;
+      }
+      entity = parent;
+    }
+    return true;
+  }
```
Comment (zhuxudong) [inline packages/core/src/mesh/SkinnedMeshRenderer.ts:338]:
```ts
private _getEntityHierarchyPath(rootEntity: Entity, searchEntity: Entity, inversePath: number[]): boolean {
    inversePath.length = 0;
    while (searchEntity !== rootEntity) {
        if (!searchEntity.parent) {
            return false;
        }
        inversePath.push(searchEntity.siblingIndex);
        searchEntity = searchEntity.parent;
    }
    return true;
}
```

File: packages/core/src/mesh/SkinnedMeshRenderer.ts [hunk 5]
```diff
@@ -244,8 +253,25 @@ export class SkinnedMeshRenderer extends MeshRenderer {
   /**
    * @internal
    */
-  override _cloneTo(target: SkinnedMeshRenderer): void {
-    super._cloneTo(target);
+  override _cloneTo(target: SkinnedMeshRenderer, srcRoot: Entity, targetRoot: Entity): void {
```
Comment (zhuxudong) [inline packages/core/src/mesh/SkinnedMeshRenderer.ts:256]: if(this._jointEntities) ->  if(this._bones)

File: packages/core/src/mesh/Skin.ts [hunk 6]
```diff
@@ -9,6 +9,10 @@ export class Skin extends EngineObject {
   public inverseBindMatrices: Matrix[];
   public joints: string[];
   public skeleton: string;
```
Comment (zhuxudong) [inline packages/core/src/mesh/Skin.ts:11]: `joints` and `skeleton` are not used now

File: packages/loader/src/gltf/parser/GLTFSkinParser.ts [hunk 7]
```diff
@@ -8,47 +8,44 @@ export class GLTFSkinParser extends GLTFParser {
...
+          const bone = entities[joints[i]];
+          skin.bones[i] = bone;
+          skin.joints[i] = bone.name;
+
+          // Get skeleton
           if (skeleton !== undefined) {
-            skin.skeleton = entities[skeleton].name;
+            const rootBone = entities[skeleton];
+            skin.rootBone = rootBone;
+            skin.skeleton = rootBone.name;
```
Comment (zhuxudong) [inline packages/loader/src/gltf/parser/GLTFSkinParser.ts:44]: delete

File: packages/loader/src/gltf/parser/GLTFSkinParser.ts [hunk 8]
```diff
@@ -8,47 +8,44 @@ export class GLTFSkinParser extends GLTFParser {
...
           } else {
             const rootBone = this._findSkeletonRootBone(joints, entities);
             if (rootBone) {
+              skin.rootBone = rootBone;
               skin.skeleton = rootBone.name;
```
Comment (zhuxudong) [inline packages/loader/src/gltf/parser/GLTFSkinParser.ts:49]: delete

File: tests/src/loader/GLTFLoader.test.ts [hunk 9]
```diff
@@ -420,5 +420,12 @@ describe("glTF Loader test", function () {
     const renderer = entities[1].getComponent(SkinnedMeshRenderer);
     expect(renderer).to.exist;
     expect(renderer.blendShapeWeights).to.deep.include([1, 1]);
+    // expect(renderer.rootBone).to.equal();
+    // expect(renderer.bones).to.equal();
```
Comment (zhuxudong) [inline tests/src/loader/GLTFLoader.test.ts:424]: delete

File: tests/src/loader/GLTFLoader.test.ts [hunk 10]
```diff
@@ -420,5 +420,12 @@ describe("glTF Loader test", function () {
...
+
+    glTFResource.destroy();
+    expect(glTFResource.materials).to.be.null;
+    expect(glTFResource.textures).to.be.null;
+    expect(glTFResource.entities).to.be.null;
```
Comment (zhuxudong) [inline tests/src/loader/GLTFLoader.test.ts:429]: merge error, delete it

File: packages/core/src/mesh/SkinnedMeshRenderer.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/mesh/SkinnedMeshRenderer.ts:185]: done

File: packages/core/src/mesh/SkinnedMeshRenderer.ts (hunk 2 above)
Comment (GuoLei1990) [reply packages/core/src/mesh/SkinnedMeshRenderer.ts:344]: done

File: packages/core/src/mesh/SkinnedMeshRenderer.ts (hunk 4 above)
Comment (GuoLei1990) [reply packages/core/src/mesh/SkinnedMeshRenderer.ts:338]: done

File: packages/core/src/mesh/Skin.ts (hunk 6 above)
Comment (GuoLei1990) [reply packages/core/src/mesh/Skin.ts:11]: Keep readonly

File: tests/src/loader/GLTFLoader.test.ts (hunk 9 above)
Comment (GuoLei1990) [reply tests/src/loader/GLTFLoader.test.ts:424]: done

File: packages/core/src/mesh/SkinnedMeshRenderer.ts (hunk 5 above)
Comment (GuoLei1990) [reply packages/core/src/mesh/SkinnedMeshRenderer.ts:256]: done

File: tests/src/loader/GLTFLoader.test.ts (hunk 10 above)
Comment (GuoLei1990) [reply tests/src/loader/GLTFLoader.test.ts:429]: done

**Review by zhuxudong — APPROVED** [review]

**Pull request merged**